### ISSUE TITLE
Add type hints

### ISSUE THREAD

**Issue by fzumstein** [xlwings/xlwings#2201]
* First priority is the public API under https://github.com/xlwings/xlwings/blob/main/xlwings/main.py
* Happy to drop Python 3.7 support for this as in June is EOL, but it should be compatible with Python 3.8+, there are some type hint enhancements that are only compatible with Python 3.10+ 
* There are tools such as: https://github.com/Instagram/MonkeyType
* Maybe CoPilot can also help

**Comment by Jeroendevr** [comment]
Note to self

- [ ] One particular helpful improvement is to be made to the `.option()` [type](https://docs.xlwings.org/en/latest/api/range.html#xlwings.Range.options). Right now IDE's do not give intellisense suggestions.

**Comment by Jeroendevr** [comment]
@fzumstein Wanted to make a start with type hinting but there is one thing I don't fully understand. Could you elaborate on the use of `impl` variabale. 

```Python

class App:
 
    """Cut docs"""

    def __init__(self, visible=None, spec=None, add_book=True, impl=None):
        if impl is None:
            self.impl = engines.active.apps.add(
                spec=spec, add_book=add_book, visible=visible
            ).impl
            if visible or visible is None:
                self.visible = True
        else:
            self.impl = impl
            if visible:
                self.visible = True
        self._pid = self.pid

```

**Comment by fzumstein** [comment]
these are the objects of the respective engine, e.g. a pywin32 object on Windows or an appscript object on macOS.

**Comment by fzumstein** [comment]
Let me try again: it's a platform-specific implementation of the respective [base class](https://github.com/xlwings/xlwings/blob/main/xlwings/base_classes.py). In your example, it would be `_xlwindows.App` for Windows and `_xlmac.App` for macOS.

**Comment by fzumstein** [comment]
Using MonkeyType would work something like this:

```
pip install MonkeyType
pip install pytest-monkeytype
pytest ./tests/test_app.py --monkeytype-output=./monkeytype.sqlite3
monkeytype list-modules
monkeytype stub xlwings.main > stubs.py
monkeytype apply xlwings.main
```
Error messages such as

```
monkeytype apply xlwings.main                
ERROR: Failed applying stub with libcst:
Could not resolve a unique qualified name for type App at 39:28. Candidate names were: ['xlwings._xlmac.App', 'App']
```

Refer to the position in the stubs.py file.

There's a PR https://github.com/Instagram/MonkeyType/pull/302 that may fix this, not sure.

**Comment by fzumstein** [comment]
done in #2685
